### DISCUSSION TITLE
Determine field value offsets?

### DISCUSSION THREAD

**Discussion by dsorber** [simdjson/simdjson#1505] [Q&A]
First off, I just stumbled upon simdjson this afternoon and it looks awesome! 

I have a perhaps somewhat unique application where rather than parsing a JSON document and extracting field values, I instead need to extract field value offsets (start and end ideally).  After looking through the examples and API documentation I don't see any mention of position or offsets.  Is it possible to extract field value offset information with simdjson?

**Comment by lemire** [comment]
Can you elaborate ? Maybe with an example and some JSON ?

(It saves time when we understanding exactly what everyone means.)

**Reply by dsorber** [reply]
Certainly, I should have included that. Here is a very simple example, consider this JSON:

`{"a": 378, "b": null, "c": "foobar"}`

Is there are way to determine that the value of key/field `"a"` begins at byte offset 7 (the `'3'` character) and ends at byte offset 8 (the `'8'` character)? Likewise that the value of key/field `"c"` begins at byte offset 28 (the `'f'` character) and ends at byte offset 33 (the `'r'` character)? 

For the one application I mentioned, I'm interested in finding the position with in a buffer (i.e. offset) where a key's value occurs.

**Reply by lemire** [reply]
For the keys, we do provide a pointer to the start of the key... With the On Demand API, when processing the keys, we often just return a raw pointer to the buffer, see...

https://github.com/simdjson/simdjson/blob/master/include/simdjson/generic/ondemand/raw_json_string.h

So you have a memory pointer right at the beginning of the key. The `raw_json_string` is not aware of its length in our current API but that's a minor issue.

For the strings values in general, we escape them to a separate buffer where they get normalized to unescaped UTF-8, so we don't track the location of the string in the input.

**Comment by jkeiser** [comment]
While iterating through JSON, you can also call get_raw_token() on any value, which gives you a start offset and a "maximum length" (which includes the entire string / double in the case of scalars, as well as any trailing whitespace). Since it returns a pointer into the original buffer, you can easily get the offset from that.

**Reply by lemire** [reply]
+1

**Reply by jkeiser** [reply]
Oh ... also of note, if you do it this way you can skip parsing of the actual values, too (or just get their type).

**Reply by dsorber** [reply]
I assume you meant [raw_json_token()](https://simdjson.org/api/0.9.0/classsimdjson_1_1_s_i_m_d_j_s_o_n___i_m_p_l_e_m_e_n_t_a_t_i_o_n_1_1ondemand_1_1value.html#ab3f9b17e853c3b7ed8cd7ca9bb9b05c6) ? I think I can make that work, I'll give it a try hopefully later today.  Thanks for your response.
